Add more properties controlling service reference code generation

dougbu opened the pull request:
- `$(OpenApiGenerateBeforeCompile)` controls if targets run before compile targets
  - #4924
  - also correct multiple invocations when project has multiple target frameworks
- `$(OpenApiBuildReferencedProjects)` controls whether `@(OpenApiProjectReference)` items build automatically
  - #6582

also:
- add symbols for Microsoft.Extensions.ApiDescription.Client task assembly
  - #10508
- unconditionally run `OpenApiGetDocuments` target in referenced projects
  - corrects compilation in design-time builds
  - no longer uses `@(ProjectReferenceWithConfiguration)`; referenced project chooses all property values

nits:
- don't remove `$(TargetFrameworks)` when invoking `OpenApiGetDocuments` target (not necessary)
- consolidate into a single `$(GenerateOpenApiReferenceCodeDependsOn)` property
- rename task assembly and namespaces in Microsoft.Extensions.ApiDesription.Client to match the project
- allow `OpenApiGetDocuments` targets to run in parallel if `$(BuildInParallel)` is enabled
- remove `$(OpenApiDefaultOutputDirectory)` normalization; never concatenated with anything else
- remove dangling `GenerateServiceFileReferenceCode` names

dougbu committed:
Add more properties controlling service reference code generation
- `$(OpenApiGenerateBeforeCompile)` controls if targets run before compile targets
  - #4924
  - also correct multiple invocations when project has multiple target frameworks
- `$(OpenApiBuildReferencedProjects)` controls whether `@(OpenApiProjectReference)` items build automatically
  - #6582

also:
- add symbols for Microsoft.Extensions.ApiDescription.Client task assembly
  - #10508
- unconditionally run `OpenApiGetDocuments` target in referenced projects
  - corrects compilation in design-time builds
  - no longer uses `@(ProjectReferenceWithConfiguration)`; referenced project chooses all property values

nits:
- don't remove `$(TargetFrameworks)` when invoking `OpenApiGetDocuments` target (not necessary)
- consolidate into a single `$(GenerateOpenApiReferenceCodeDependsOn)` property
- rename task assembly and namespaces in Microsoft.Extensions.ApiDesription.Client to match the project
- allow `OpenApiGetDocuments` targets to run in parallel if `$(BuildInParallel)` is enabled
- remove `$(OpenApiDefaultOutputDirectory)` normalization; never concatenated with anything else
- remove dangling `GenerateServiceFileReferenceCode` names

dougbu commented on src/Mvc/Extensions.ApiDescription.Client/src/build/Microsoft.Extensions.ApiDescription.Client.targets:
```diff
@@ -30,19 +27,16 @@
+        Projects="@(OpenApiProjectReference)"
         RebaseOutputs="true"
-        RemoveProperties="%(ProjectReferenceWithConfiguration.GlobalPropertiesToRemove);TargetFrameworks;RuntimeIdentifier">
+        RemoveProperties="Configuration;Platform;RuntimeIdentifier;TargetFramework;TargetFrameworks">
```
@rainersigwald is removing these properties sufficient?

And, am I missing something important in the shift from `@(ProjectReferenceWithConfiguration)` to using `@(OpenApiProjectReference)` items directly? (Without this, design-time builds would still have been broken and `'$(OpenApiBuildReferencedProjects)' == 'false'` would not work.)

rainersigwald replied:
I'd definitely want to keep something like `%(OpenApiProjectReference.GlobalPropertiesToRemove)` in there as an escape hatch. This list looks ok to me, but there's always a risk of things changing (like when multitargeting was added) or user projects having their own custom dimensions.

dougbu replied:
Great suggestion! I added this to #4923 because I have other bits to do short-term.

dougbu commented:
@natemcmaster I'd of course appreciate any input you have on the MSBuild changes

dougbu commented on src/Mvc/Extensions.ApiDescription.Client/src/build/Microsoft.Extensions.ApiDescription.Client.props:
```diff
@@ -1,35 +1,58 @@
   <!--
     Settings users may update as they see fit.
   -->
   <PropertyGroup>
```
@glennc do you have any input on the names of these user-visible properties?

pranavkm commented on src/Mvc/Extensions.ApiDescription.Client/src/build/Microsoft.Extensions.ApiDescription.Client.targets:
```diff
@@ -30,19 +27,16 @@
-  <Target Name="_CreateOpenApiReferenceItemsForOpenApiProjectReferences"
-      AfterTargets="ResolveProjectReferences"
-      Condition="'$(DesignTimeBuild)' != 'true' AND '$(BuildingProject)' == 'true'">
+  <Target Name="_CreateOpenApiReferenceItemsForOpenApiProjectReferences" DependsOnTargets="ResolveProjectReferences">
```
Should this have `Condition="@(OpenApiProjectReference->Count()) != 0"`?

dougbu replied:
It really doesn't matter because everything in this target no-ops when there's nothing worth doing. Let me know if you feel differently

rynowak replied:
Simple is better, I agree with Doug.

pranavkm commented on src/Mvc/Extensions.ApiDescription.Client/src/build/Microsoft.Extensions.ApiDescription.Client.props:
```diff
@@ -1,35 +1,58 @@
+      If 'false', the GenerateOpenApiReferenceCode target is not part of the build (by default) but can run when
+      explicitly referenced. That is, the target may be invoked from the command line or tied in through another target.
+    -->
+    <OpenApiGenerateBeforeCompile Condition="'$(OpenApiGenerateBeforeCompile)' == ''">true</OpenApiGenerateBeforeCompile>
```
`OpenApiGenerateOnBuild`? That's similar to `RazorCompileOnBuild`. Also removes having this be tied to `BeforeCompile` which is essentially an implementation detail.

dougbu replied:
This property is specifically about moving execution elsewhere in the build. So, I like the current name. But, I defer to @glennc here.

glennc replied:
OpenApiGenerateCodeOnBuild

pranavkm commented on src/Mvc/Extensions.ApiDescription.Client/src/buildMultiTargeting/Microsoft.Extensions.ApiDescription.Client.targets:
```diff
@@ -1,9 +1,14 @@
   </Target>
+
+  <Target Name="_TieInGenerateOpenApiReferenceCode"
+      BeforeTargets="BeforeCompile"
```
Should this be `CoreCompile`. Would that allow users to do user things ™️ in `BeforeCompile`?

dougbu replied:
`CoreCompile` is late because it comes after the MSBuild targets cache the list of references and source files involved in the compilation

pranavkm reviewed (approved):
Looks good to me. I wouldn't block on any of my comments.

dougbu committed:
Rename zee properties
- make it clear these properties control code generation
- use "OnBuild" for consistency with properties elsewhere in ASP.NET Core
- remove "Default" from a comple of property names because commaents are pretty clear
- rename a few targets for consistency with the new properties
  - word "Reference" didn't add much

rynowak commented on src/Mvc/Extensions.ApiDescription.Client/src/build/Microsoft.Extensions.ApiDescription.Client.props:
```diff
@@ -1,35 +1,58 @@
+        Condition="'$(MSBuildRuntimeType)' == 'Core'">netstandard2.0</_ApiDescriptionClientAssemblyTarget>
+    <_ApiDescriptionClientAssemblyTarget
+        Condition="'$(MSBuildRuntimeType)' != 'Core'">net461</_ApiDescriptionClientAssemblyTarget>
+    <_ApiDescriptionClientAssemblyPath>$(MSBuildThisFileDirectory)/../tasks/$(_ApiDescriptionClientAssemblyTarget)/Microsoft.Extensions.ApiDescription.Client.dll</_ApiDescriptionClientAssemblyPath>
```
Is this really relevant anymore? MSBuild supports netstandard2.0 everywhere .NET Core runs.

rainersigwald replied:
If the assembly has any dependencies that aren't in netstandard _as implemented in .NET 4.7.2_, I'd leave this in. Netstandard task assemblies are probably much more viable now that VS is on 4.7.2, but they've historically been a problem and splitting into core and full implementations has been required.

rynowak replied:
I'm mentioning it because we've been doing ns2.0 in Razor for a while now with no problems. 

Since this is new code it should be fine as long as we test it in VS right?

rainersigwald replied:
Yeah, as long as it works in VS, command line msbuild.exe, and command line msbuild.exe from a Build Tools install of VS, you should be good to go.

dougbu replied:
I'll file an issue deferring this cleanup and testing in to Preview 7

dougbu replied:
I've added this to #4923 since the extra bits do no harm. This is all an implementation detail and it'll take time to do the necessary testing&hellip;

rynowak commented on eng/ProjectReferences.props:
```diff
@@ -8,7 +8,7 @@
     <ProjectReferenceProvider Include="Microsoft.Web.Xdt.Extensions" ProjectPath="$(RepoRoot)src\SiteExtensions\Microsoft.Web.Xdt.Extensions\src\Microsoft.Web.Xdt.Extensions.csproj"  />
     <ProjectReferenceProvider Include="Microsoft.AspNetCore.DeveloperCertificates.XPlat" ProjectPath="$(RepoRoot)src\Tools\FirstRunCertGenerator\src\Microsoft.AspNetCore.DeveloperCertificates.XPlat.csproj"  />
-    <ProjectReferenceProvider Include="Microsoft.Extensions.ApiDescription.Tasks" ProjectPath="$(RepoRoot)src\Mvc\Extensions.ApiDescription.Client\src\Microsoft.Extensions.ApiDescription.Client.csproj"  />
+    <ProjectReferenceProvider Include="Microsoft.Extensions.ApiDescription.Client" ProjectPath="$(RepoRoot)src\Mvc\Extensions.ApiDescription.Client\src\Microsoft.Extensions.ApiDescription.Client.csproj"  />
```
I might be missing some necessary context here, but I don't really get the point of renaming an assembly containing tasks to not be suffixed with `.Tasks`. 

Is the metapoint here that we want to separate the tasks used for client-gen from the tasks used inside server-projects?

dougbu replied:
That was my original point but I ended up not needing tasks in the server project and decided I preferred using the same name everywhere.

rynowak commented on src/Mvc/Extensions.ApiDescription.Client/src/build/Microsoft.Extensions.ApiDescription.Client.props:
```diff
@@ -1,35 +1,58 @@
-  <UsingTask TaskName="GetCurrentItems" AssemblyFile="$(_ApiDescriptionTasksAssemblyPath)" />
-  <UsingTask TaskName="GetFileReferenceMetadata" AssemblyFile="$(_ApiDescriptionTasksAssemblyPath)" />
+  <UsingTask TaskName="GetCurrentItems" AssemblyFile="$(_ApiDescriptionClientAssemblyPath)" />
+  <UsingTask TaskName="GetFileReferenceMetadata" AssemblyFile="$(_ApiDescriptionClientAssemblyPath)" />
```
random observation: `GetCurrentItems` is a pretty generic name given that `item` is a common term in MSBuild. We might want to choose more specific names. I don't think `GetFileReferenceMetadata` is bad.

dougbu replied:
I added the suggestion to #4923

rynowak commented on src/Mvc/Extensions.ApiDescription.Client/src/build/Microsoft.Extensions.ApiDescription.Client.props:
```diff
@@ -1,35 +1,58 @@
+    <!--
+      Options added to the code generator command line by default. Provides the default %(Options) metadata of
+      @(OpenApiReference) and @(OpenApiProjectReference) items.
+    -->
```
I'm always really happy to see quality docs in MSBuild 👍

rynowak commented on src/Mvc/Extensions.ApiDescription.Client/src/build/Microsoft.Extensions.ApiDescription.Client.props:
```diff
@@ -1,35 +1,58 @@
+
+    <!--
+      If 'true' (the default), generate code for @(OpenApiReference) and @(OpenApiProjectReference) items during
+      design-time builds. Otherwise, generate code only during a full build.
```
If you're going with the name `OpenApiGenerateCodeOnBuild` would this become `OpenApiGenerateCodeOnDesignTimeBuild`?

dougbu replied:
Thought about it but the sense is a bit different. @glennc and I decided this one was fine as-is. Could of course change it if users disagree.

rynowak commented on src/Mvc/Extensions.ApiDescription.Client/src/build/Microsoft.Extensions.ApiDescription.Client.targets:
```diff
@@ -85,10 +79,9 @@
       Outputs="%(OutputPath)">
-    <MSBuild BuildInParallel="$(BuildInParallel)"
-        Projects="$(MSBuildProjectFullPath)"
+    <MSBuild Projects="$(MSBuildProjectFullPath)"
```
Is this idea with this call that we only want to call this once even with multi-targeting? If that's the case should we have a comment?

Basically I'm not sure why this is a separate invocation of MSBuild, a comment would help.

dougbu replied:
Yes, this is about running things once. And, sure, I'll add a comment

rynowak reviewed (approved):
I left some minor comments here but nothing blocking.

dougbu committed:
Add comment to client tie-in target

dougbu commented:
:up::date: to add the comment @rynowak suggested. Also rebased on 'master' but deferred most other suggestions to #4923

dougbu committed:
Correct target name in buildMultiTargeting/Microsoft.Extesions.ApiDescription.Client.targets
